text/0000-contract-metadata.md

```diff
@@ -0,0 +1,117 @@
+- Proposal Name: `contract_metadata`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- NEP PR: [nearprotocol/neps#0003](https://github.com/nearprotocol/NEPs/pull/3)
+
+# Summary
+[summary]: #summary
+
+This NEP introduces contract metadata, which summarizes the content of a given contract in json format.
+Contract metadata allows developers to easily list the methods in a contract and potentially display them to users to
+provide transparency.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no direct way for a developer to programmatically list the methods of a contract that is deployed on chain
+because the contract code stored on chain is the compiled wasm code, where the methods name and parameters are already mangled.
+As a result, if developers want to display the methods of a contract to an user of the app to provide transparency,
+especially in the case of financial apps, they have no choice but to hardcode them in the front end, which is suboptimal in many ways.
+Furthermore, if developers want to get the list of methods in some contract for some downstream task like data analysis,
+they have no way of doing so. Contract metadata aims to solve the aforementioned problems by providing a convenient way 
+for developers to list the methods of contracts.
+
+# Guide-level explanation
```

Comment:
This NEP should be rephrased in language-agnostic way. We now have two contract languages: AssemblyScript and Rust, since Rust will be the primary language for financial instruments it makes it the first-class citizen on par with AssemblyScript.

Reply:
I don't see where in the description there is language-specific information. In examples, yes, but it is not possible to have real examples without being language-specific. Same with reference-level explanations. If you want to add rust examples, feel free to do so.

Reply:
If you are proposing a change to the system, make sure it also includes all the components of the system. Rust is one of such components.

Reply:
Sure but I am not familiar with how rust contracts work, so I think it would be better if someone who's familiar with how rust bindings are implemented write the corresponding part.

Reply:
The exact format of how we encode argument types and return types into JSON will affect Rust. Because if we come up with some non-standard method not supported by standard Rust JSON serializer/deserializer then it will be a huge problem and a waste of time tweaking Rust JSON serializer/deserializer to understand it.

Reply:
> Sure but I am not familiar with how rust contracts work, so I think it would be better if someone who's familiar with how rust bindings are implemented write the corresponding part.

I suggest you familiarize yourself. Rust is going to be a huge part of our system, all important contracts are going to be written on it. I cannot be the only person knowing how it works.

Reply:
It was my understanding from the latest call (with @vgrichina) that typescript would be the focus, if rust is the main language you plan to use, would it serve us better to write our contracts in rust (and if so, is there a benefit to rust vs typescript from this POV?)

Reply:
Rust contracts-writing UX is optimized for safety while typescript contracts-writing UX is optimized for convenience. It is a trade-off that the contract developers have to make.